tests/unit/test_instance.py

```diff
@@ -285,3 +281,162 @@ async def test_Instance_close(instance: Instance) -> None:
     assert (instance._current.done() or instance._current.cancelled()) is True
     assert instance._next.cancelled() is True
     assert instance._client_session.closed is True
+
+
+@pytest.mark.asyncio
+@patch("google.cloud.sql.connector.instance._get_ephemeral")
+@patch("google.cloud.sql.connector.instance._get_metadata")
+async def test_perform_refresh(
+    mock_metadata: Dict,
+    mock_ephemeral: str,
+    fake_credentials: Credentials,
+    event_loop: asyncio.AbstractEventLoop,
+) -> None:
+    """
+    Test that _perform_refresh returns valid InstanceMetadata object.
+    """
+    # generate keys and certs for test
+    cert, priv_key = generate_cert("my-project", "my-instance")
+    server_ca_cert = self_signed_cert(cert, priv_key)
+    keys = asyncio.create_task(get_keys())
+    _, client_key = await keys
+    ephemeral_cert = client_key_signed_cert(cert, priv_key, client_key)
```

Comment:
Rather than have to do this a bunch of times, would it be better to make a "FakeProxyServer" object or similar tha does it in one location?

For example, see what we did in the Go connector: https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/blob/e29f8c13583d3ddb65ebec907ccdd300643efd82/internal/mock/cloudsql.go#L146

It's a single object creation but guarantees we create a fake backend, including the API requests and proxy server.